File: order-service/src/main/java/com/intexsoft/webshop/orderservice/service/impl/OrderServiceImpl.java
package com.intexsoft.webshop.orderservice.service.impl;

import com.intexsoft.webshop.orderservice.dto.detail.DetailCreateDto;
import com.intexsoft.webshop.orderservice.dto.order.OrderCreateDto;
import com.intexsoft.webshop.orderservice.dto.order.OrderDto;
import com.intexsoft.webshop.orderservice.mapper.OrderMapper;
import com.intexsoft.webshop.orderservice.model.Order;
import com.intexsoft.webshop.orderservice.model.Status;
import com.intexsoft.webshop.orderservice.model.enums.StatusName;
import com.intexsoft.webshop.orderservice.producer.OrderEventProducer;
import com.intexsoft.webshop.orderservice.repository.OrderRepository;
import com.intexsoft.webshop.orderservice.service.OrderService;
import com.intexsoft.webshop.orderservice.util.JsonUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class OrderServiceImpl implements OrderService {
    private final OrderRepository orderRepository;
    private final OrderEventProducer orderEventProducer;
    private final OrderMapper orderMapper;

    @Override
    @Transactional
    public OrderDto saveOrder(OrderCreateDto orderCreateDto) {
        log.info("IN: trying to save a new order = {}", JsonUtils.getAsString(orderCreateDto));
        Status initStatus = new Status(StatusName.INITIALIZED);
        BigDecimal totalPrice = calculateOrderTotalPrice(orderCreateDto.getDetailCreateDtos());
        Order savedOrder = orderRepository.save(orderMapper.toOrder(orderCreateDto, initStatus, totalPrice));
        log.info("OUT: the order saved successfully. The saved order = {}",
                JsonUtils.getAsString(savedOrder));
        orderEventProducer.produceOrderInitializedEvent(orderMapper.toOrderInitializedEvent(savedOrder));
        return orderMapper.toOrderDto(savedOrder);
    }

    private BigDecimal calculateOrderTotalPrice(List<DetailCreateDto> detailCreateDtos) {
        return detailCreateDtos.stream()
                .map(detailCreateDto -> {
                    BigDecimal productPrice = detailCreateDto.getProductPrice();
                    Short quantity = detailCreateDto.getQuantity();
                    return productPrice.multiply(BigDecimal.valueOf(quantity));
                })
                .reduce(BigDecimal.ZERO, BigDecimal::add)
                .setScale(2, RoundingMode.HALF_UP);
    }
}

File: order-service-kt/src/main/kotlin/com/intexsoft/webshop/orderservicekt/service/impl/OrderServiceImpl.kt
package com.intexsoft.webshop.orderservicekt.service.impl

import com.intexsoft.webshop.orderservicekt.dto.detail.DetailCreateDto
import com.intexsoft.webshop.orderservicekt.dto.order.OrderCreateDto
import com.intexsoft.webshop.orderservicekt.dto.order.OrderDto
import com.intexsoft.webshop.orderservicekt.log
import com.intexsoft.webshop.orderservicekt.mapper.DetailMapper
import com.intexsoft.webshop.orderservicekt.mapper.OrderMapper
import com.intexsoft.webshop.orderservicekt.model.Order
import com.intexsoft.webshop.orderservicekt.model.Status
import com.intexsoft.webshop.orderservicekt.producer.OrderEventProducer
import com.intexsoft.webshop.orderservicekt.repository.OrderRepository
import com.intexsoft.webshop.orderservicekt.service.OrderService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.math.RoundingMode

@Service
class OrderServiceImpl(
    private val orderRepository: OrderRepository,
    private val orderEventProducer: OrderEventProducer,
    private val orderMapper: OrderMapper,
    private val detailMapper: DetailMapper
) : OrderService {

    @Transactional
    override fun saveOrder(orderCreateDto: OrderCreateDto): OrderDto {
        log.info("IN: trying to save a new order = $orderCreateDto")
        val detailCreateDtos = orderCreateDto.detailCreateDtos
        val totalPrice = calculateOrderTotalPrice(detailCreateDtos)
        val newOrder = orderMapper.toOrder(orderCreateDto, totalPrice)
        newOrder.addStatus(Status())
        detailCreateDtos.forEach { newOrder.addDetail(detailMapper.toDetail(it)) }
        val savedOrder: Order = orderRepository.save(newOrder)
        log.info("OUT: the order saved successfully. The saved order = $savedOrder")
        orderEventProducer.produceOrderInitializedEvent(orderMapper.toOrderInitializedEvent(savedOrder))
        return orderMapper.toOrderDto(savedOrder)
    }

    private fun calculateOrderTotalPrice(detailCreateDtos: List<DetailCreateDto>): BigDecimal {
        return detailCreateDtos
            .map { it.productPrice.multiply(BigDecimal(it.quantity.toInt())) }
            .fold(BigDecimal.ZERO) { acc: BigDecimal, inc: BigDecimal? -> acc.add(inc) }
            .setScale(2, RoundingMode.HALF_UP)
    }
}
